README
This perl5 jitter is super-simple, and modeled after B::CC

The compiled perl5 optree is a linked list in memory in non-execution order,
wide-spread jumps. Additionally the calls are indirect with a shared
libperl. The jitter properly aligns the run-time calls in linear linked-list
"exec" order, so that the CPU can prefetch the next instructions, and it inlines
some simple ops.  op_next targets (returned by false conditions) are favored
over op_other and other targets.

IT DOES NOT WORK YET!
It does only work for simple functions, no branching ops yet.

Faster jitted execution path without runops loop, selected with -MJit or later,
when stable with perl -j.

All ops are unrolled in execution order for the CPU cache,
prefetching is the main advantage of this function. The perl5 runloop has 
no chance to get cached at all.
For < 5.13 the ASYNC check is only done when necessary.

For now only implemented for x86 and amd64/x86_64 with certain 
hardcoded my_perl offsets when threaded.

C pseudocode

x86 not-threaded, PL_op in eax, PL_sig_pending temp in ecx

prolog:
	55                   	push   %ebp
	89 e5                	mov    %esp,%ebp
	53                    	push   %rbx 
call:
	e8 xx xx xx xx		call   $PL_op->op_ppaddr #relative
save_plop:
	a3 xx xx xx xx       	mov    %eax,$PL_op

dispatch_getsig:
	8b 0d xx xx xx xx	mov    $PL_sig_pending,%ecx
dispatch:
	85 c9                	test   %ecx,%ecx
	74 06                	je     nextcall
	e8 xx xx xx xx          call   *Perl_despatch_signals #relative
epilog:
	b8 00 00 00 00       	mov    $0x0,%eax 	# clean PL_op
	5b                   	pop    %rbx 
	5d               	pop    %ebp
	c3                   	ret

If op maybranch (see also Opcodes-0.04), jit the other op branches, 
check PL_op before with after and branch to label of other op.

Porting
I created the asm with cc_main and cc_main_nt, see Makefile for objdump and cc_harness 
rules for gcc assembly.

ASM links

http://www.lxhp.in-berlin.de/lhplinks.html
http://blogs.msdn.com/freik/archive/2005/03/17/398200.aspx
http://msdn.microsoft.com/en-us/library/7kcdt6fy.aspx
http://asm.sourceforge.net//resources.html
http://www.intel.com/design/itanium/manuals/iiasdmanual.htm
http://www.heyrick.co.uk/assembler/qfinder.html

HL jitters

parrot
luajit
psyco / pypy
tracemonkey
ruby
clisp

JIT libs

lightning - c macros only
libjit - c lib
llvm - compiler framework + lib


Changes
v0.05 2012-04-02 rurban

cleanup ptr conversions
fixed amd64 non-threaded

v0.04_10 2010-12-13 rurban

remove unused returnother, LOGOP wo/ entertry it is

v0.04_09 2010-12-03 rurban

%rbx on amd64 seems not be save across function calls.
  store op->next on 0(%rsp) ("the local stack") - not.
  store it at -8(%rbp)
protect %edx (used in dispatch handling)
prevent dryrun also from recursive cycles, but with a different HV
 Fix jmp byte overflow, use word size now.
re-work and document MAYBRANCH ops, sync with new Opcodes-0.10
  use simple GOTO as default (searches at run-time in targets, does not work yet)
first ENTERSUB attempt, but CvSTART detection would be much easier
  from the parser
more JIT_CHAIN helper macros
added libdisasm
properly NOP end of code (alignment)
rename help dir to asm

v0.04_08 2010-11-22 rurban

work on non-local jumps: GOTO and loops, new test 4.t
memalign fixes for openbsd: use size == pagesize when memalign==malloc
fixed maybranch for the logop else case (3.t test 1/2).
use high-level enter 8,0 paired to leave.
HV* otherops to check against recursion loops.
better tests: fix for spaces in path, 5.6: -MBlib workaround
improved make clean

v0.04_07 2010-11-18 rurban

fixes for <5.10: CopLABEL, push_arg1

v0.04_06 2010-11-18 rurban

fixed posix_memalign (freebsd)
added preliminary support for non-local jumps (goto + loops), not working though

v0.04_05 2010-11-16 rurban

fixed amd64thr: my_perl is initially at %eax
fixed threaded -Ds debstack calls, -Dt fails
fixed HAVE_DISPATCH on all 4 platforms
work on mybranch: move if maybranch: op = PL_op->next before call,
  without dynamic next maybranch_check 3.t even works.
simplify jit_chain code
harmonize amd64 maybranch asm (store next away at r12),
use dbg_lines macros instead of ugly #ifdef orgies

t/2.t works now mostly, t/3.t not yet working

v0.04_04 2010-11-15 rurban

enabled HAVE_DISPATCH on i386, tested ok on i386 threaded and non-threaded.
on amd64 untested.

v0.04_03 2010-11-14 rurban

fix tests for Win cmd.exe shell. Tested ok on strawberry and ActivePerl 5.8
started work on WIN64 calling conventions (no test system yet).
  rcx is not safe to use on amd64 for &PL_sig_pending
fixed solaris 10 / cc 32-bit
fixed perl5.6.2 (check old macros)

v0.04_02 2010-11-13 rurban

fixed -Dt for amd64
do not support MULTIPLICITY without ITHREADS
2.t first enter works again on i386, just not when calling -Dt (probably some register sideeffect).

v0.04_01 2010-11-12 rurban

make amd64 non-threaded work again (old absolute save_plop)
  amd64 r12 unused, only rbx and rcx. maybranch not yet refactored

forgot LOGOP JIT_CHAIN code+root, so the op->next block (true) was never emitted
1st simple profiling looks promising
  See http://blogs.perl.org/users/rurban/2010/11/performance-hacks.html
Try to overcome a missing memalign, by manually aligning the malloced code section, 
  first to pagesize, then to 4096. Fails tests on cygwin, but this has memalign.
implement -Dt debop op tracing within the jit, also -Ds debstack
add my_perl to jit_chain
rename PUSHmov to PUSHrel     
CALL_ABS changes code
add absword and revword macros, use it
add push_imm, push_imm_0 (call any functions)

skip t/3.t on no DEBUGGING perl

check old DEBUG_s..., Perl_ck_warner_d, fix pod
adjust docs (threaded works now)
add explicit abstract and license

TODO first PL_op enter definitevly missing (test 2)

v0.04 2010-11-07 rurban

Fixed i386 1.t + 2.t, threaded and non-threaded
  wrong mov %eax,(%ebx), 
  mov $mem,%ebx (forgot shifting)
  save_plop now without argument
t/3.t maybranch still fails

v0.0302 2010-11-06 rurban

Added memalign, posix_memalign probes. Got failing reports for OpenBSD.

v0.0301 2010-08-28 rurban

Oops, add missing {i386,amd64}.c code to MANIFEST

v0.03 2010-08-09 rurban

Start with amd64/x86_64
Include seperate CPU threaded/non-threaded c files
Create run-jit.c and run-jit.o with .stabs when DEBUGGING
Refactor (maybranch) into jit_chain which runs twice to 
  support branches eventually - ops which do not return op_next
seperate tests, only 1.t works

v0.02 2010-08-06 rurban

Found the i386 call bug with rafl's help - "use stepi"
  - call = e9 (using offsets!), not ff 25 (jmp far).
Fixed call near offsets and PL_op offset without USE_JMP_TABLE.
Fixed initialising PL_op
Small i386 (non-threading) programs do work now

v0.01 2010-08-05 rurban

Non-working version pushed to CPAN being on YAPC::EU. Wanted to show it in a
lightning talk.

- 2010-03-14 rurban

Created and published at http://gist.github.com/331867


asm/3-jit.c
typedef struct op OP; struct op {OP* op_next};
OP *PL_op; int *Perl_Isig_pending_ptr; void runops_jit_0 (void);
void runops_jit_0 (void){
    register OP* op; 
    register int *plop = &PL_op;
    register int *p = &Perl_Isig_pending_ptr;
    PL_op = Perl_pp_enter();
    PL_op = Perl_pp_nextstate();
    PL_op = Perl_pp_const();
    PL_op = Perl_pp_padsv();
    PL_op = Perl_pp_sassign();
    PL_op = Perl_pp_nextstate();
    if (*p)
        Perl_despatch_signals();
    PL_op = Perl_pp_padsv();
    PL_op = Perl_pp_const();
    PL_op = Perl_pp_gt();

/* maybranch_1: */
    op = PL_op->op_next; /* save 4(%ebx) (PL_op->next) into %ecx */
    PL_op = Perl_pp_cond_expr();
    if (*p)
        Perl_despatch_signals();
    /* simple case: only op->other possible */
    if (PL_op == op) /* cmp eax to ecx (prev op->next) if false */
        goto next_1;
 other_1:
    PL_op = Perl_pp_pushmark();
    PL_op = Perl_pp_const();
    PL_op = Perl_pp_print();
    goto leave_1; /* eof next_1, upper scope */
 next_1:
    PL_op = Perl_pp_enter();
    PL_op = Perl_pp_nextstate();
    if (*p)
        Perl_despatch_signals();
    PL_op = Perl_pp_leave();
 leave_1:
    PL_op = Perl_pp_leave();
    return;
}

int main() {
    runops_jit_0();
}

/* unthreaded i386:
gcc -fstack-protector 3-jit.c /usr/lib/perl5/5.10/i686-cygwin/CORE/libperl.a -lcrypt 
objdump -d a.exe > 3-jit.exedis-static

00401130 <_runops_jit_0>:
  401130:	55                   	push   %ebp
  401131:	89 e5                	mov    %esp,%ebp
  401133:	83 ec 08             	sub    $0x8,%esp
  401136:	c7 45 fc d0 90 50 00 	movl   $0x5090d0,-0x4(%ebp)
  40113d:	e8 ce 11 00 00       	call   402310 <_Perl_pp_enter>
  401142:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401147:	e8 24 31 00 00       	call   404270 <_Perl_pp_nextstate>
  40114c:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401151:	e8 aa a1 00 00       	call   40b300 <_Perl_pp_const>
  401156:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  40115b:	e8 f0 96 00 00       	call   40a850 <_Perl_pp_padsv>
  401160:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401165:	e8 86 9d 00 00       	call   40aef0 <_Perl_pp_sassign>
  40116a:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  40116f:	e8 fc 30 00 00       	call   404270 <_Perl_pp_nextstate>
  401174:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401179:	8b 55 fc             	mov    -0x4(%ebp),%edx
  40117c:	8b 02                	mov    (%edx),%eax
  40117e:	85 c0                	test   %eax,%eax
  401180:	74 05                	je     401187 <_runops_jit_0+0x57> 182+05=187
  401182:	e8 19 b8 01 00       	call   41c9a0 <_Perl_despatch_signals>
  401187:	e8 c4 96 00 00       	call   40a850 <_Perl_pp_padsv>
  40118c:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401191:	e8 6a a1 00 00       	call   40b300 <_Perl_pp_const>
  401196:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  40119b:	e8 f0 3a 01 00       	call   414c90 <_Perl_pp_gt>
  4011a0:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011a5:	a1 c0 75 54 00       	mov    0x5475c0,%eax
  4011aa:	89 45 f8             	mov    %eax,-0x8(%ebp)
  4011ad:	e8 ce 31 00 00       	call   404380 <_Perl_pp_cond_expr>
  4011b2:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011b7:	8b 55 fc             	mov    -0x4(%ebp),%edx
  4011ba:	8b 02                	mov    (%edx),%eax
  4011bc:	85 c0                	test   %eax,%eax
  4011be:	74 05                	je     4011c5 <_runops_jit_0+0x95> 1c0+05=1c5
  4011c0:	e8 db b7 01 00       	call   41c9a0 <_Perl_despatch_signals>
  4011c5:	a1 c0 75 54 00       	mov    0x5475c0,%eax
  4011ca:	3b 45 f8             	cmp    -0x8(%ebp),%eax
  4011cd:	74 20                	je     4011ef <_runops_jit_0+0xbf>
  4011cf:	e8 4c 57 00 00       	call   406920 <_Perl_pp_pushmark>
  4011d4:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011d9:	e8 22 a1 00 00       	call   40b300 <_Perl_pp_const>
  4011de:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011e3:	e8 38 86 00 00       	call   409820 <_Perl_pp_print>
  4011e8:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011ed:	eb 2c                	jmp    40121b <_runops_jit_0+0xeb> 1ef+2c=21b
  4011ef:	e8 1c 11 00 00       	call   402310 <_Perl_pp_enter>
  4011f4:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  4011f9:	e8 72 30 00 00       	call   404270 <_Perl_pp_nextstate>
  4011fe:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401203:	8b 55 fc             	mov    -0x4(%ebp),%edx
  401206:	8b 02                	mov    (%edx),%eax
  401208:	85 c0                	test   %eax,%eax
  40120a:	74 05                	je     401211 <_runops_jit_0+0xe1>
  40120c:	e8 8f b7 01 00       	call   41c9a0 <_Perl_despatch_signals>
  401211:	e8 4a 1b 00 00       	call   402d60 <_Perl_pp_leave>
  401216:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  40121b:	e8 40 1b 00 00       	call   402d60 <_Perl_pp_leave>
  401220:	a3 c0 75 54 00       	mov    %eax,0x5475c0
  401225:	c9                   	leave  
  401226:	c3                   	ret    

amd64:
gcc -fstack-protector 3-jit.c /usr/lib/perl5/5.13.3/i686-nothreads-debug-linux/CORE/libperl.a -lcrypt -lm
objdump -d a.out > 3-jit.exedis-static64

0000000000404fd4 <runops_jit_0>:
  404fd4:	55                   	push   %rbp
  404fd5:	48 89 e5             	mov    %rsp,%rbp
  404fd8:	41 54                	push   %r12
  404fda:	53                   	push   %rbx
  404fdb:	bb 10 61 9d 00       	mov    $0x9d6110,%ebx
  404fe0:	b8 00 00 00 00       	mov    $0x0,%eax
  404fe5:	e8 86 a7 03 00       	callq  43f770 <Perl_pp_enter>
  404fea:	48 98                	cltq   
  404fec:	48 89 05 15 11 5d 00 	mov    %rax,0x5d1115(%rip)        # 9d6108 <PL_op>
  404ff3:	b8 00 00 00 00       	mov    $0x0,%eax
  404ff8:	e8 f9 89 02 00       	callq  42d9f6 <Perl_pp_nextstate>
  404ffd:	48 98                	cltq   
  404fff:	48 89 05 02 11 5d 00 	mov    %rax,0x5d1102(%rip)        # 9d6108 <PL_op>
  405006:	b8 00 00 00 00       	mov    $0x0,%eax
  40500b:	e8 6c 89 02 00       	callq  42d97c <Perl_pp_const>
  405010:	48 98                	cltq   
  405012:	48 89 05 ef 10 5d 00 	mov    %rax,0x5d10ef(%rip)        # 9d6108 <PL_op>
  405019:	b8 00 00 00 00       	mov    $0x0,%eax
  40501e:	e8 98 bd 02 00       	callq  430dbb <Perl_pp_padsv>
  405023:	48 98                	cltq   
  405025:	48 89 05 dc 10 5d 00 	mov    %rax,0x5d10dc(%rip)        # 9d6108 <PL_op>
  40502c:	b8 00 00 00 00       	mov    $0x0,%eax
  405031:	e8 2d 93 02 00       	callq  42e363 <Perl_pp_sassign>
  405036:	48 98                	cltq   
  405038:	48 89 05 c9 10 5d 00 	mov    %rax,0x5d10c9(%rip)        # 9d6108 <PL_op>
  40503f:	b8 00 00 00 00       	mov    $0x0,%eax
  405044:	e8 ad 89 02 00       	callq  42d9f6 <Perl_pp_nextstate>
  405049:	48 98                	cltq   
  40504b:	48 89 05 b6 10 5d 00 	mov    %rax,0x5d10b6(%rip)        # 9d6108 <PL_op>
  405052:	8b 03                	mov    (%rbx),%eax
  405054:	85 c0                	test   %eax,%eax
  405056:	74 0a                	je     405062 <runops_jit_0+0x8e>
  405058:	b8 00 00 00 00       	mov    $0x0,%eax
  40505d:	e8 a4 6f 00 00       	callq  40c006 <Perl_despatch_signals>
  405062:	b8 00 00 00 00       	mov    $0x0,%eax
  405067:	e8 4f bd 02 00       	callq  430dbb <Perl_pp_padsv>
  40506c:	48 98                	cltq   
  40506e:	48 89 05 93 10 5d 00 	mov    %rax,0x5d1093(%rip)        # 9d6108 <PL_op>
  405075:	b8 00 00 00 00       	mov    $0x0,%eax
  40507a:	e8 fd 88 02 00       	callq  42d97c <Perl_pp_const>
  40507f:	48 98                	cltq   
  405081:	48 89 05 80 10 5d 00 	mov    %rax,0x5d1080(%rip)        # 9d6108 <PL_op>
  405088:	b8 00 00 00 00       	mov    $0x0,%eax
  40508d:	e8 d0 7c 0b 00       	callq  4bcd62 <Perl_pp_gt>
  405092:	48 98                	cltq   
  405094:	48 89 05 6d 10 5d 00 	mov    %rax,0x5d106d(%rip)        # 9d6108 <PL_op>
  40509b:	48 8b 05 66 10 5d 00 	mov    0x5d1066(%rip),%rax        # 9d6108 <PL_op>
  4050a2:	4c 8b 20             	mov    (%rax),%r12
  4050a5:	b8 00 00 00 00       	mov    $0x0,%eax
  4050aa:	e8 38 a7 02 00       	callq  42f7e7 <Perl_pp_cond_expr>
  4050af:	48 98                	cltq   
  4050b1:	48 89 05 50 10 5d 00 	mov    %rax,0x5d1050(%rip)        # 9d6108 <PL_op>
  4050b8:	8b 03                	mov    (%rbx),%eax
  4050ba:	85 c0                	test   %eax,%eax
  4050bc:	74 0a                	je     4050c8 <runops_jit_0+0xf4>
  4050be:	b8 00 00 00 00       	mov    $0x0,%eax
  4050c3:	e8 3e 6f 00 00       	callq  40c006 <Perl_despatch_signals>
  4050c8:	48 8b 05 39 10 5d 00 	mov    0x5d1039(%rip),%rax        # 9d6108 <PL_op>
  4050cf:	4c 39 e0             	cmp    %r12,%rax
  4050d2:	74 3b                	je     40510f <runops_jit_0+0x13b>
  4050d4:	b8 00 00 00 00       	mov    $0x0,%eax
  4050d9:	e8 70 8c 02 00       	callq  42dd4e <Perl_pp_pushmark>
  4050de:	48 98                	cltq   
  4050e0:	48 89 05 21 10 5d 00 	mov    %rax,0x5d1021(%rip)        # 9d6108 <PL_op>
  4050e7:	b8 00 00 00 00       	mov    $0x0,%eax
  4050ec:	e8 8b 88 02 00       	callq  42d97c <Perl_pp_const>
  4050f1:	48 98                	cltq   
  4050f3:	48 89 05 0e 10 5d 00 	mov    %rax,0x5d100e(%rip)        # 9d6108 <PL_op>
  4050fa:	b8 00 00 00 00       	mov    $0x0,%eax
  4050ff:	e8 36 08 03 00       	callq  43593a <Perl_pp_print>
  405104:	48 98                	cltq   
  405106:	48 89 05 fb 0f 5d 00 	mov    %rax,0x5d0ffb(%rip)        # 9d6108 <PL_op>
  40510d:	eb 4a                	jmp    405159 <runops_jit_0+0x185> # leave_1
  40510f:	90                   	nop
  405110:	b8 00 00 00 00       	mov    $0x0,%eax
  405115:	e8 56 a6 03 00       	callq  43f770 <Perl_pp_enter>
  40511a:	48 98                	cltq   
  40511c:	48 89 05 e5 0f 5d 00 	mov    %rax,0x5d0fe5(%rip)        # 9d6108 <PL_op>
  405123:	b8 00 00 00 00       	mov    $0x0,%eax
  405128:	e8 c9 88 02 00       	callq  42d9f6 <Perl_pp_nextstate>
  40512d:	48 98                	cltq   
  40512f:	48 89 05 d2 0f 5d 00 	mov    %rax,0x5d0fd2(%rip)        # 9d6108 <PL_op>
  405136:	8b 03                	mov    (%rbx),%eax
  405138:	85 c0                	test   %eax,%eax
  40513a:	74 0a                	je     405146 <runops_jit_0+0x172>
  40513c:	b8 00 00 00 00       	mov    $0x0,%eax
  405141:	e8 c0 6e 00 00       	callq  40c006 <Perl_despatch_signals>
  405146:	b8 00 00 00 00       	mov    $0x0,%eax
  40514b:	e8 f3 b0 03 00       	callq  440243 <Perl_pp_leave>
  405150:	48 98                	cltq   
  405152:	48 89 05 af 0f 5d 00 	mov    %rax,0x5d0faf(%rip)        # 9d6108 <PL_op>
leave_1:
  405159:	b8 00 00 00 00       	mov    $0x0,%eax
  40515e:	e8 e0 b0 03 00       	callq  440243 <Perl_pp_leave>
  405163:	48 98                	cltq   
  405165:	48 89 05 9c 0f 5d 00 	mov    %rax,0x5d0f9c(%rip)        # 9d6108 <PL_op>
  40516c:	5b                   	pop    %rbx
  40516d:	41 5c                	pop    %r12
  40516f:	c9                   	leaveq 
  405170:	c3                   	retq   


*/
